### PR TITLE
feat: Update the signatures for variants and options to follow convention

### PR DESCRIPTION
**Pull request by sradevski**

### CODE REVIEW

File: packages/core-flows/src/handlers/product/update-product-variants.ts [hunk 1]
```diff
@@ -27,7 +27,7 @@ export async function updateProductVariants({
   }
```
Comment (adrien2p) [inline packages/core-flows/src/handlers/product/update-product-variants.ts:27]: todo: general, missing change set

File: packages/core-flows/src/product/steps/update-product-options.ts [hunk 2]
```diff
@@ -39,11 +40,10 @@ export const updateProductOptionsStep = createStep(
...
-    // TODO: We need to update the module's signature
-    // await service.upsertOptions(
-    //   prevData.map((r) => ({
-    //     ...r,
-    //   }))
-    // )
+    await service.upsertOptions(
+      prevData.map((r) => ({
+        ...r,
+      }))
```
Comment (adrien2p) [inline packages/core-flows/src/product/steps/update-product-options.ts]: Q: do we need the shallow copy here?

File: packages/core-flows/src/product/steps/update-product-variants.ts [hunk 3]
```diff
@@ -39,11 +40,10 @@ export const updateProductVariantsStep = createStep(
...
-    // TODO: We need to update the module's signature
-    // await service.upsertVariants(
-    //   prevData.map((r) => ({
-    //     ...r,
-    //   }))
-    // )
+    await service.upsertVariants(
+      prevData.map((r) => ({
+        ...r,
+      }))
```
Comment (adrien2p) [inline packages/core-flows/src/product/steps/update-product-variants.ts]: same question

File: packages/product/src/services/product-module-service.ts [hunk 4]
```diff
@@ -197,58 +198,158 @@ export default class ProductModuleService<
...
-        )[0]
-      })
+    const createdVariants = await this.baseRepository_.serialize<
+      ProductTypes.ProductVariantDTO[]
+    >(variants, { populate: true })
+
+    return Array.isArray(data) ? createdVariants : createdVariants[0]
+  }
+
+  async createVariants_(
```
Comment (adrien2p) [inline packages/product/src/services/product-module-service.ts]: todo: missing InjectTransactionManager

File: packages/product/src/services/product-module-service.ts [hunk 5]
```diff
@@ -197,58 +198,158 @@ export default class ProductModuleService<
...
+    if (data.some((v) => !v.product_id)) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_DATA,
+        "Tried to create variants without specifying a product_id"
+      )
+    }
+
+    const productOptions = await this.productOptionService_.list(
+      {
+        product_id: uniqBy(data, "product_id").map(
```
Comment (adrien2p) [inline packages/product/src/services/product-module-service.ts]:
```suggestion
        product_id: [...new Set<string>(data.map(d => d.product_id))]
```

Comment (adrien2p) [reply packages/product/src/services/product-module-service.ts]: not sure this util uniqBy is required, also, a single loop is necessary, which also simplify the readability and cyclomatic complexity, even if we dont really care here

File: packages/product/src/services/product-module-service.ts [hunk 6]
```diff
@@ -197,58 +198,158 @@ export default class ProductModuleService<
...
+    data:
+      | ProductTypes.UpsertProductVariantDTO[]
+      | ProductTypes.UpsertProductVariantDTO,
+    @MedusaContext() sharedContext: Context = {}
+  ): Promise<
+    ProductTypes.ProductVariantDTO[] | ProductTypes.ProductVariantDTO
+  > {
+    const input = Array.isArray(data) ? data : [data]
+    const forUpdate = input.filter(
+      (variant): variant is UpdateProductVariantInput => !!variant.id
```
Comment (adrien2p) [inline packages/product/src/services/product-module-service.ts:281]: info: can't wait the new incoming version, which will remove the need to do that ahah

File: packages/product/src/services/product-module-service.ts [hunk 7]
```diff
@@ -197,58 +198,158 @@ export default class ProductModuleService<
...
+    )
+    const forCreate = input.filter(
+      (variant): variant is ProductTypes.CreateProductVariantDTO => !variant.id
+    )
```
Comment (adrien2p) [inline packages/product/src/services/product-module-service.ts:285]:
```suggestion
const forUpdate: UpdateProductVariantInput[] = []
const forCreate: CreateProductVariantDTO[] = []
input.forEach(variant => {
  if (!!variant.id) {
    forUpdate.push(variant as UpdateProductVariantInput)
    return
  }
  
  forCreate.push(variant as CreateProductVariantDTO)
})
```

File: packages/product/src/services/product-module-service.ts [hunk 8]
```diff
@@ -197,58 +198,158 @@ export default class ProductModuleService<
...
-  ): Promise<ProductTypes.ProductVariantDTO[]> {
-    const productVariants = await this.updateVariants_(data, sharedContext)
+  ): Promise<
+    ProductTypes.ProductVariantDTO[] | ProductTypes.ProductVariantDTO
+  > {
+    let normalizedInput: UpdateProductVariantInput[] = []
+    if (isString(idOrSelector)) {
+      normalizedInput = [{ id: idOrSelector, ...data }]
+    } else {
+      const variants = await this.productVariantService_.list(
```
Comment (adrien2p) [inline packages/product/src/services/product-module-service.ts:328]: q: is it really necessary to list in order to get the id only? knowing that the internal update method accept an array of selector/data? We can improve later, but in theory we have many queries to fetch the same data here and there.

File: packages/product/src/services/product-module-service.ts [hunk 9]
```diff
@@ -197,58 +198,158 @@ export default class ProductModuleService<
...
-    >(productVariants)
+  async upsertVariants(
+    data: ProductTypes.UpsertProductVariantDTO[],
+    sharedContext?: Context
+  ): Promise<ProductTypes.ProductVariantDTO[]>
+  async upsertVariants(
+    data: ProductTypes.UpsertProductVariantDTO,
+    sharedContext?: Context
+  ): Promise<ProductTypes.ProductVariantDTO>
+  @InjectTransactionManager("baseRepository_")
```
Comment (adrien2p) [inline packages/product/src/services/product-module-service.ts:270]: nit: add some spec

File: packages/product/src/services/product-module-service.ts [hunk 10]
```diff
@@ -427,11 +528,44 @@ export default class ProductModuleService<
...
+    const options = await this.createOptions_(input, sharedContext)
+
+    const createdOptions = await this.baseRepository_.serialize<
+      ProductTypes.ProductOptionDTO[]
+    >(options, { populate: true })
+
+    return Array.isArray(data) ? createdOptions : createdOptions[0]
+  }
+
+  async createOptions_(
```
Comment (adrien2p) [inline packages/product/src/services/product-module-service.ts]: todo: InjectTransactionManager

Comment (adrien2p) [inline packages/product/src/services/product-module-service.ts]: todo: protected?

File: packages/product/src/services/product-module-service.ts (hunk 4 above)
Comment (adrien2p) [inline packages/product/src/services/product-module-service.ts]: todo: protected

File: packages/product/src/services/product-module-service.ts [hunk 11]
```diff
@@ -441,21 +575,101 @@ export default class ProductModuleService<
...
+
+  async upsertOptions(
+    data: ProductTypes.UpsertProductOptionDTO[],
+    sharedContext?: Context
+  ): Promise<ProductTypes.ProductOptionDTO[]>
+  async upsertOptions(
+    data: ProductTypes.UpsertProductOptionDTO,
+    sharedContext?: Context
+  ): Promise<ProductTypes.ProductOptionDTO>
+  @InjectTransactionManager("baseRepository_")
```
Comment (adrien2p) [inline packages/product/src/services/product-module-service.ts:593]: nit: some space

File: packages/product/src/services/product-module-service.ts [hunk 12]
```diff
@@ -441,21 +575,101 @@ export default class ProductModuleService<
...
+      | ProductTypes.UpsertProductOptionDTO,
+    @MedusaContext() sharedContext: Context = {}
+  ): Promise<ProductTypes.ProductOptionDTO[] | ProductTypes.ProductOptionDTO> {
+    const input = Array.isArray(data) ? data : [data]
+    const forUpdate = input.filter(
+      (option): option is UpdateProductOptionInput => !!option.id
+    )
+    const forCreate = input.filter(
+      (option): option is ProductTypes.CreateProductOptionDTO => !option.id
+    )
```
Comment (adrien2p) [inline packages/product/src/services/product-module-service.ts:606]: same suggestion as the previous one, feel free to not do it if you want

File: packages/core-flows/src/handlers/product/update-product-variants.ts (hunk 1 above)
Comment (sradevski) [reply packages/core-flows/src/handlers/product/update-product-variants.ts:27]: Ah right, we decided to start adding changesets even for incomplete v2 work.

File: packages/core-flows/src/product/steps/update-product-options.ts (hunk 2 above)
Comment (sradevski) [reply packages/core-flows/src/product/steps/update-product-options.ts]: I don't think so no 👍

File: packages/product/src/services/product-module-service.ts (hunk 7 above)
Comment (sradevski) [reply packages/product/src/services/product-module-service.ts:285]: To be honest, I prefer the readability of the current solution, even if it loops twice (it should be insignificant, and if we ever see that's where the performance issue lies, we can optimize)

File: packages/product/src/services/product-module-service.ts (hunk 8 above)
Comment (sradevski) [reply packages/product/src/services/product-module-service.ts:328]: @adrien2p this is the pattern we followed elsewhere, and it's about having the `update_` methods operating on top of an update object with IDs, and not relying on the selector. I am fine either way to be honest, but for now I won't touch it and we can revisit all modules later

File: packages/core-flows/src/handlers/product/update-product-variants.ts (hunk 1 above)
Comment (adrien2p) [reply packages/core-flows/src/handlers/product/update-product-variants.ts:27]: The product module is already released and accessible, thats why but maybe not? @olivermrbl

File: packages/product/src/services/product-module-service.ts (hunk 12 above)
Comment (sradevski) [reply packages/product/src/services/product-module-service.ts:606]: Same comment, I'd leave it for now 👍

File: packages/product/src/services/product-module-service.ts (hunk 8 above)
Comment (adrien2p) [reply packages/product/src/services/product-module-service.ts:328]: I believe it was a mistake in the other modules as people didn't know that the internal method accepted selectors, fine to keep it as it is for now and we can come back to it at a later point. I remember pointing this to seb pr's too

File: packages/product/src/services/product-module-service.ts [hunk 13]
```diff
@@ -295,7 +395,7 @@ export default class ProductModuleService<
   }
 
   @InjectManager("baseRepository_")
-  async diffVariants_(
+  protected async diffVariants_(
```
Comment (adrien2p) [inline packages/product/src/services/product-module-service.ts:398]: todo: InjectTransactionManager here

**Review by adrien2p — APPROVED** [review]
LGTM 🔥

**Review by adrien2p — COMMENTED** [review]
LGTM, one last comment 🔥

**Pull request merged**